bigquery_schema_generator/generate_schema.py

```diff
@@ -768,27 +842,104 @@ def flatten_schema_map(
                         keep_nulls=keep_nulls,
                         sorted_schema=sorted_schema,
                         infer_mode=infer_mode,
-                        sanitize_names=sanitize_names,
+                        input_format=input_format
                     )
             elif key == 'type' and value in ['QINTEGER', 'QFLOAT', 'QBOOLEAN']:
-                # Convert QINTEGER -> INTEGER, similarly for QFLAT and QBOOLEAN.
+                # Convert QINTEGER -> INTEGER, similarly for QFLOAT and QBOOLEAN
                 new_value = value[1:]
             elif key == 'mode':
-                if infer_mode and value == 'NULLABLE' and filled:
+                # 'infer_mode' to set a field as REQUIRED is supported for only
+                # input_format = 'csv' because the header line gives us the
+                # complete list of fields to be expected in the CSV file. In
+                # JSON data files, certain fields will often be completely
+                # missing instead of being set to 'null' or "". If the field is
+                # not even present, then it becomes incredibly difficult (not
+                # impossible, but more effort than I want to expend right now)
+                # to figure out which fields are missing so that we can mark the
+                # appropriate schema entries with 'filled=False'.
+                if (infer_mode and value == 'NULLABLE' and filled
+                        and input_format == 'csv'):
```

Comment:
Ah, yes, now I understand. In the context of `flatten_schema_map()`, the `--infer_mode` will be active only if the `input_format` is CSV. So this allows us to overload the `--infer_mode` flag to also indicate, "allow REQUIRED -> NULLABLE" transition. This is subtle even for me. Can you add a comment just above this, saying something like:
```
The --infer_mode option is activated only for input_format == 'csv' in this function,
which allows us to overload the --infer_mode flag to mean that a REQUIRED
mode of an existing schema can transition to a NULLABLE mode.
```

Can we also add a test in `testdata.txt` to verify this? Something like:
```
# Verify that the --infer_mode flag for JSON input does NOT cause the schema mode to 
# become REQUIRED, even though in the following, it will mistakenly think that every field
# defined for all records, because it cannot (easily) detect the absence of a field in a record.
DATA infer_mode
{ "i" : 1, "f": 1.0, "r": {"a": 1} }
{ "f": 1.0 }
SCHEMA
[
  {
    "mode": "NULLABLE",
    "name": "f",
    "type": "FLOAT"
  },
  {
    "mode": "NULLABLE",
    "name": "i",
    "type": "INTEGER"
  },
  {
    "fields": [
      {
        "mode": "NULLABLE",
        "name": "a",
        "type": "INTEGER"
      }
    ],
    "mode": "NULLABLE",
    "name": "r",
    "type": "RECORD"
  }
]
END
```